mapiproxy/servers/default/emsmdb/emsmdbp_object.c

```diff
@@ -1882,6 +1955,121 @@ _PUBLIC_ void **emsmdbp_object_table_get_row_props(TALLOC_CTX *mem_ctx, struct e
         return data_pointers;
 }
 
+
+
+/**
+   \details This function process the hierarchy of folders recursively
+   and fill requested rows.
+
+   \param mem_ctx pointer to the memory context
+   \param emsmdbp_ctx pointer to the emsmdb context
+   \param context_object pointer to the context object
+   \param datablob pointer to the DATA blob to fill
+   \param properties pointer to the array of properties used to fill a
+   row of the table
+   \param current_fid the FolderID of the folder to open and process
+   \param remaining pointer on the remaining rows to process
+   \param count pointer on the number of rows processed to return
+
+   \note This implementation is very preliminary and lacks several use cases:
+   1. It does not handle backward read
+   2. It does not handle positioning of the folder object and rewind
+   of the hierarchy when position != 0
+
+   We currently rely on Outlook asking for 4096 rows to fill this
+   buffer in one operation.
+
+   \return MAPI_E_SUCCESS on success, otherwise MAPI error.
+*/
+_PUBLIC_ enum MAPISTATUS emsmdbp_object_table_get_recursive_row_props(TALLOC_CTX *mem_ctx, struct emsmdbp_context *emsmdbp_ctx,
+								      struct emsmdbp_object *context_object,
+								      DATA_BLOB *datablob, struct SPropTagArray *properties,
+								      uint64_t current_fid, int64_t *remaining, uint32_t *count)
+{
+	enum MAPISTATUS		retval;
+	struct emsmdbp_object	*folder = NULL;
+	struct emsmdbp_object	*table = NULL;
+	struct mapi_handles	*rec = NULL;
+	enum MAPISTATUS		*retvals;
+	void			**data_pointers = NULL;
+	uint32_t		handle = 0;
+	uint32_t		i = 0;
+	uint32_t		j;
+
+	/* Sanity checks */
+	OPENCHANGE_RETVAL_IF(!emsmdbp_ctx, MAPI_E_INVALID_PARAMETER, NULL);
+	OPENCHANGE_RETVAL_IF(!context_object, MAPI_E_INVALID_PARAMETER, NULL);
+	OPENCHANGE_RETVAL_IF(!remaining, MAPI_E_INVALID_PARAMETER, NULL);
+	OPENCHANGE_RETVAL_IF(!*remaining, MAPI_E_INVALID_PARAMETER, NULL);
+	OPENCHANGE_RETVAL_IF(!count, MAPI_E_INVALID_PARAMETER, NULL);
+
+	/* open current folder */
+	if (current_fid) {
+		retval = emsmdbp_object_open_folder_by_fid(mem_ctx, emsmdbp_ctx, context_object, current_fid, &folder);
+		if (retval) {
+			return MAPI_E_INVALID_OBJECT;
+		}
+		retval = mapi_handles_add(emsmdbp_ctx->handles_ctx, handle, &rec);
+		if (retval) {
+			talloc_free(folder);
+			return MAPI_E_INVALID_OBJECT;
+		}
+		table = emsmdbp_folder_open_table(rec, folder, MAPISTORE_FOLDER_TABLE, rec->handle);
```

Comment:
cleanup required upon failure + free object in further calls if error encountered